feat(account): Implement the Ability to Add Recovery Providers by Governance Action

enddynayn opened the pull request:
This pull request introduces new functionality to the MSA (Message Source Account) pallet by adding support for managing Recovery Providers, including their approval and removal via governance. It also updates related documentation, benchmarking, and tests to reflect these changes.

### New Features for Recovery Providers:
* Added a new storage type `RecoveryProviders` to store approved Recovery Providers.
* Introduced two new extrinsics:
  - `approve_recovery_provider`: Allows governance to approve a Recovery Provider, emitting the `RecoveryProviderApproved` event.
  - `remove_recovery_provider`: Allows governance to remove a Recovery Provider, emitting the `RecoveryProviderRemoved` event.

### Updates to Configuration and Benchmarks:
* Added `RecoveryProviderApprovalOrigin` to the configuration, defining the origin required for approving recovery providers
* Added benchmarks for the new extrinsics `approve_recovery_provider` and `remove_recovery_provider`.

### Documentation Enhancements:
* Updated the README to include details about the new extrinsics and state queries related to Recovery Providers.

### Testing Improvements:
* Added governance tests to verify the functionality of approving and removing Recovery Providers, including edge cases like unauthorized access and non-existent providers.

### Weight Adjustments:
* Updated the weight functions to include the new extrinsics `approve_recovery_provider` and `remove_recovery_provider`.

# Checklist
- [x] Updated Pallet Readme?
- [ ] Updated js/api-augment for Custom RPC APIs?
- [x] Design doc(s) updated?
- [x] Unit Tests added?
- [ ] e2e Tests added? [out of scope for this PR, tracking issue #2491]
- [x] Benchmarks added?
- [x] Spec version incremented?

mattheworris committed:
feat: add secure recovery key system design documentation for Frequency MSA

mattheworris committed:
refactor: update section headings and improve structure in recovery system documentation

mattheworris committed:
feat: enhance recovery process with new validation and event handling for recovery keys

mattheworris committed:
refactor: update participant names in account creation sequence diagram for clarity

mattheworris committed:
refactor: update terminology from recovery key to recovery secret for consistency in documentation

mattheworris committed:
refactor: update terminology and improve clarity in recovery system documentation

enddynayn committed:
feat(account): add Recovery Provider support (#2469)

Introduce support for designating and managing Recovery Providers,
allowing only trusted entities to perform account recovery actions.

Recovery Providers must be approved through governance and must already
be registered as a regular provider before they can take on the recovery
role.

issue-2464

mattheworris committed:
refactor: standardize terminology and improve clarity in recovery system documentation

mattheworris committed:
feat(wip): updates to data structures; remove secrets from events;

enddynayn committed:
Update weights

mattheworris committed:
refactor: update terminology from Recovery Secret to Recovery Hash for consistency in documentation; address PR comments

mattheworris committed:
refactor: clarify Recovery Secret generation process and update terminology for consistency

mattheworris committed:
Merge branch 'main' of github.com:frequency-chain/frequency into feat/recovery-design-development

mattheworris committed:
refactor: update terminology from Recovery Hash to Recovery Commitment for consistency in documentation

mattheworris committed:
refactor: update pseudocode

mattheworris committed:
Merge branch 'feat/recovery-design-development' of github.com:frequency-chain/frequency into feat/account-recovery-development

mattheworris committed:
refactor: enhance recovery process documentation and clarify off-chain validation steps

mattheworris committed:
fix: correct sender in recovery email sequence diagram

mattheworris committed:
wip(feat): implement add_recovery_commitment functionality with associated tests and types

mattheworris committed:
feat: implement add_recovery_commitment functionality in WeightInfo and associated tests

mattheworris committed:
chore: bump spec_version to 167 for runtime updates

mattheworris committed:
Update weights

mattheworris committed:
refactor: update weight calculations for add_recovery_commitment and adjust related test values

wilwade committed:
Recovery System JS SDK (#2472)

# Goal
The goal of this PR is to generate the JS SDK that makes it easy to
generate and use the recovery hashes and commitments

Closes #2466 

# Discussion

- Added CI
- Added Recovery SDK

---------

Co-authored-by: Matthew Orris <1466844+mattheworris@users.noreply.github.com>

mattheworris committed:
feat: implement test that adds add_recovery_commitment to capacity transactions

mattheworris committed:
feat: add recovery_commitment field to RecoveryCommitmentAdded event and update related tests

mattheworris committed:
feat: enhance RecoveryCommitmentPayload with discriminant and update related functions

mattheworris committed:
fix: update concatenation notation and clarify Recovery Commitment handling in recovery system documentation

mattheworris committed:
fix: clarify terminology in recovery system documentation and update provider references

mattheworris committed:
Merge branch 'main' of github.com:frequency-chain/frequency into feat/add-recovery-commitment-2465

mattheworris committed:
feat: update RecoveryCommitmentPayload encoding to include chain ID

mattheworris committed:
feat: add recovery-sdk to E2E workflow and installation steps

mattheworris committed:
fix: update deny.toml to include new advisory and allow additional GitHub organization

mattheworris committed:
fix: bump spec_version to 168 for runtime compatibility

mattheworris committed:
fix: update runtime version for recovery commitment functions to 168

wilwade committed:
Merge remote-tracking branch 'origin/main' into feat/add-recovery-commitment-2465

# Conflicts:
#	e2e/package-lock.json

mattheworris committed:
Merge branch 'main' of github.com:frequency-chain/frequency into feat/add-recovery-commitment-2465

mattheworris committed:
feat: implement recovery commitment functionality and tests

mattheworris committed:
Update js/recovery-sdk/CONTRIBUTING.md

Co-authored-by: Wil Wade <wil.wade@projectliberty.io>

mattheworris committed:
bug: address pr comments

mattheworris committed:
Merge branch 'feat/add-recovery-commitment-2465' of github.com:frequency-chain/frequency into feat/add-recovery-commitment-2465

mattheworris committed:
feat: add RecoveryCommitmentPayload interface and related functionality to ethereum-utils

mattheworris committed:
wip(feat): enhance recovery commitment tests and update RecoveryCommitmentData interface

mattheworris committed:
wip(refactor): rename RecoveryCommitmentData to RecoveryCommitmentPayload for clarity

mattheworris committed:
fix: update recovery commitment payload type from Uint8Array to string in multiple files; grab signature from metamask

mattheworris committed:
test: update recovery commitment payload in signature tests and metamask integration

wilwade committed:
Recovery SDK tweaks and Updated MSA Pallet Readme (#2479)

Recovery SDK is now at 100% test coverage.

wilwade committed:
Tweak wording around "seed" for recovery generation (#2480)

I think this covers @JoeCap08055's review comments

mattheworris committed:
refactor: update RecoveryCommitmentPayload to use HexString and adjust related tests

mattheworris committed:
bug(ci): lint updates

mattheworris committed:
test: add case for recovery commitment with not yet valid signature

mattheworris committed:
Apply suggestions from code review

Co-authored-by: Wil Wade <wil.wade@projectliberty.io>

mattheworris committed:
refactor: simplify recovery commitment generation in tests

mattheworris committed:
feat: include recovery commitment in AccountRecovered event

mattheworris committed:
undo: unnecesary name change

aramikm committed:
fixed signatures (#2482)

# Goal
The goal of this PR is <!-- insert goal here -->

Closes <!-- issue # -->

# Discussion

- <!-- List discussion items -->

# Checklist
- [ ] Updated Pallet Readme?
- [ ] Updated js/api-augment for Custom RPC APIs?
- [ ] Design doc(s) updated?
- [ ] Unit Tests added?
- [ ] e2e Tests added?
- [ ] Benchmarks added?
- [ ] Spec version incremented?

mattheworris committed:
bug: fix eth signatures and various updates

mattheworris committed:
Merge branch 'main' of github.com:frequency-chain/frequency into feat/add-recovery-commitment-2465

mattheworris committed:
chore: update js dependencies

mattheworris committed:
fix: lint

mattheworris committed:
test: add Ethereum signed recovery commitment test case

mattheworris committed:
docs: add setup instructions for Metamask testing

mattheworris committed:
Update .github/workflows/verify-pr-commit.yml

Co-authored-by: Aramik <aramikm@gmail.com>

mattheworris committed:
Merge branch 'feat/add-recovery-commitment-2465' of github.com:frequency-chain/frequency into feat/account-recovery-development

mattheworris committed:
feat: update README and runtime version for recovery provider functionality

mattheworris committed:
bug: fix merge error

mattheworris committed:
Merge branch 'main' of github.com:frequency-chain/frequency into feat/account-recovery-development

mattheworris commented on pallets/msa/src/lib.rs:
```diff
@@ -140,6 +140,11 @@ pub mod pallet {
 		type CreateProviderViaGovernanceOrigin: EnsureOrigin<Self::RuntimeOrigin>;
 
+		/// The origin that is allowed to approve recovery providers
+		type RecoveryProviderApprovalOrigin: EnsureOrigin<
+			<Self as frame_system::Config>::RuntimeOrigin,
+		>;
```
nit: match syntax for `CreateProviderApprovalOrigin`
```suggestion
		type RecoveryProviderApprovalOrigin: EnsureOrigin<Self::RuntimeOrigin>;
```

mattheworris commented:
> > Does a Recovery Provider also need to be a Provider? This implies that Recovery Providers can be delegation targets; not sure if we want that.
> 
> Hmm... @JoeCap08055, You are saying that a Recovery Service could just do Recovery without being a Provider. I can see that. Here are the main reasons I think to keep them as Providers:
> 
> 1. Providers have access to Capacity. It is expected that these recovery services would use capacity
> 2. Providers (will) have additional on chain information about them published: [Design Doc: Provider Contexts #2311](https://github.com/frequency-chain/frequency/pull/2311)
> 
> I think those are the two main reasons. It also simplifies the tooling and the path. (Become a Provider, then request to be a Recovery Provider rather than: Become a provider if you want to use capacity which can happen at anytime, also become a recovery service)

What is the incentive to be a Recovery Provider, if you are not already a Provider?
I had some brief discussions with @enddynayn about the economic incentives and we may need to explore them further. 

However, it seemed that currently the only incentive to be an RP was to provide the additional recovery service to a Provider's users.

JoeCap08055 commented:
> > Does a Recovery Provider also need to be a Provider? This implies that Recovery Providers can be delegation targets; not sure if we want that.
> 
> Hmm... @JoeCap08055, You are saying that a Recovery Service could just do Recovery without being a Provider. I can see that. Here are the main reasons I think to keep them as Providers:
> 
> 1. Providers have access to Capacity. It is expected that these recovery services would use capacity
> 2. Providers (will) have additional on chain information about them published: [Design Doc: Provider Contexts #2311](https://github.com/frequency-chain/frequency/pull/2311)
> 
> I think those are the two main reasons. It also simplifies the tooling and the path. (Become a Provider, then request to be a Recovery Provider rather than: Become a provider if you want to use capacity which can happen at anytime, also become a recovery service)

Well, then, if being a Recovery Provider requires that you first become a Provider, we're missing some checks in this PR, I think.

Also, I thought we had an issue somewhere to track a design discussion surrounding the potential separation of the concepts of app/delegation target vs. Capacity user. I can't find the issue, but if it exists we should make a note there, and if not... do we want to create one?

JoeCap08055 commented:
TL;DR on preceding thread here:

* Since we don't have any near-term plans to separate the concept of "Capacity user" from "app/delegation target" (currently the conflation of these two attributes are what "being a Provider" means), a Recovery Provider *must* be a Provider (because they need to use Capacity)
* Therefore, we don't need a separate `RecoveryProviderId` type alias
* While a valid use case exists for a Provider to be ONLY a Recovery Provider, this is easily achieved by the RP simply not requesting any delegations or performing any other on-chain services within its app

mattheworris committed:
Merge branch 'main' of github.com:frequency-chain/frequency into feat/account-recovery-development

mattheworris committed:
Update pallets/msa/src/lib.rs

mattheworris reviewed (approved):
- Updated branch
- Added updates to README.md
- Read through changes
- Executed `make ci-local`

wilwade commented:
> Do we need an ability for a Provider to request to be a Recovery Provider? Or some other way to propose that a Provider be approved as a Recovery Provider? Or will that just be a manual governance proposal that someone creates?

I was imagining this would just be manual governance for now. Unlike a normal provider where the number would be very high, the number of recovery providers is expected to be small.

Also the "request" setup was only required for Providers as the Providers and approved via council. Recovery Providers have to submit to a full referendum

shannonwells reviewed (approved):
Only non-blocking comments, LGTM!

mattheworris commented on pallets/msa/src/lib.rs:
```diff
@@ -179,6 +182,13 @@ pub mod pallet {
 		OptionQuery,
 	>;
 
+	/// Storage type for approved recovery providers
+	/// - Key: Provider MSA Id
+	/// - Value: [`bool`]
+	#[pallet::storage]
+	pub type RecoveryProviders<T: Config> =
+		StorageMap<_, Twox64Concat, ProviderId, bool, OptionQuery>;
+
```
- Added note/question to #2269

mattheworris commented on pallets/msa/src/lib.rs:
```diff
@@ -1054,10 +1074,76 @@ pub mod pallet {
+				Error::<T>::ProviderNotRegistered
+			);
+
+			RecoveryProviders::<T>::insert(ProviderId(provider_msa_id), true);
+
+			Self::deposit_event(Event::RecoveryProviderApproved {
+				provider_id: ProviderId(provider_msa_id),
+			});
```
- Implemented check for existing approval and updated tests to reflect the change.

mattheworris commented on pallets/msa/src/tests/governance_tests.rs:
```diff
@@ -133,3 +135,106 @@ fn propose_to_be_provider_long_name_should_fail() {
+}
+
+#[test]
+fn remove_nonexistent_recovery_provider_should_succeed() {
```
- Test added to ensure that no events are emitted on the 2nd approval and the state is unaffected.

mattheworris committed:
Enhance recovery provider approval logic and tests

- Prevent duplicate approval of recovery providers in the pallet logic.
- Update governance tests to reflect changes in recovery provider approval flow.

mattheworris committed:
feat: update design doc

JoeCap08055 reviewed (approved):
My questions addressed; approved.
